sympy/assumptions/cnf.py

```diff
@@ -0,0 +1,373 @@
+
+    def __eq__(self, other):
+        return self.args == other.args
+
+    def __str__(self):
+        s = '('+' & '.join([str(arg) for arg in self.args])+')'
+        return s
+
+    __repr__ = __str__
+
+
+def to_NNF(expr):
+    """
+    Generates the Negation Normal Form of any boolean expression in terms
+    of AND , OR and Literal objects.
+    """
+    if not isinstance(expr, BooleanFunction):
+        return Literal(expr)
+
+    klass = type(expr).__name__
+    if klass == 'Not':
+        arg = expr.args[0]
+        tmp = to_NNF(arg)  # Strategy: negate the NNF of expr
+        return ~tmp
+
+    if klass == 'Or':
+        return OR(*[to_NNF(x) for x in Or.make_args(expr)])
+
+    if klass == 'And':
+        return AND(*[to_NNF(x) for x in And.make_args(expr)])
+
+    if klass == 'Nand':
+        tmp = AND(*[to_NNF(x) for x in expr.args])
+        return ~tmp
+
+    if klass == 'Nor':
+        tmp = OR(*[to_NNF(x) for x in expr.args])
+        return ~tmp
+
+    if klass == 'Xor':
+        cnfs = []
+        for i in range(0, len(expr.args) + 1, 2):
+            for neg in combinations(expr.args, i):
+                clause = [~to_NNF(s) if s in neg else to_NNF(s)
+                          for s in expr.args]
+                cnfs.append(OR(*clause))
+        return AND(*cnfs)
+
+    if klass == 'Xnor':
+        cnfs = []
+        for i in range(0, len(expr.args) + 1, 2):
+            for neg in combinations(expr.args, i):
+                clause = [~to_NNF(s) if s in neg else to_NNF(s)
+                          for s in expr.args]
+                cnfs.append(OR(*clause))
+        return ~AND(*cnfs)
+
+    if klass == 'Implies':
+        L, R = map(to_NNF, expr.args)
+        return OR(~L, R)
+
+    if klass == 'Equivalent':
+        cnfs = []
+        for a, b in zip(expr.args, expr.args[1:]):
+            a = to_NNF(a)
+            b = to_NNF(b)
+            cnfs.append(OR(~a, b))
+        a = to_NNF(expr.args[-1])
+        b = to_NNF(expr.args[0])
+        cnfs.append(OR(~a, b))
+        return AND(*cnfs)
+
+    if klass == 'ITE':
+        L, M, R = map(CNF.to_CNF, expr.args)
+        return AND(OR(~L, M), OR(L, R))
+
+
+def distribute_AND_over_OR(expr):
+    """
+    Distributes AND over OR in the NNF expression.
+    Returns the result( Conjunctive Normal Form of expression)
+    as a CNF object.
+    """
+    if not isinstance(expr, (AND, OR)):
+        tmp = set()
+        tmp.add(frozenset((expr,)))
+        return CNF(tmp)
+
+    klass = type(expr).__name__
+    if klass == 'OR':
+        return CNF.all_or(*[distribute_AND_over_OR(arg)
+                            for arg in expr._args])
+
+    if klass == 'AND':
+        return CNF.all_and(*[distribute_AND_over_OR(arg)
+                             for arg in expr._args])
+
+
+class CNF(object):
+    """
+    Class to represent CNF of a Boolean expression.
+    Consists of set of clauses, which themselves are stored as
+    frozenset of Literal objects.
```

Comment:
```suggestion
    set of Literal objects.
```

Reply:
Oh I see `clauses` is a set and its elements are frozensets.

Reply:
A set can't have unhashable types, so `frozenset` is used. I think `frozenset` should be kept.